Improve documentation of the preview mode (#10168)
astral-sh/ruff: docs/preview.md

Ruff includes an opt-in preview mode to provide an opportunity for community feedback and increase confidence that
changes are a net-benefit before enabling them for everyone.

> Preview mode enables a collection of new lint rules, fixes, formatter style and other changes that are considered experimental or unstable.

Reviewer: ```suggestion
Preview mode enables a collection of unstable features such as new lint rules and fixes, formatter style changes, interface updates, and more. Warnings about deprecated features may turn into errors when using preview mode.
```

> Simply enabling preview mode does not imply that all the new preview features/changes will be turned on. Some, like the rules that are in preview, need to be turned on explicitly (see the [rules section](#using-rules-that-are-in-preview). Others, like the formatter style or default values changes will be turned on by simply enabling the preview mode.

Reviewer: I find this a little confusing. We have a flag that determines if preview rules require "explicit" selection but by default they do not i.e. "implicit" selection of the rule via a prefix is sufficient. I'd be much more direct here:

```suggestion
Enabling preview mode does not select all preview rules by default. See the [rules section](#using-rules-that-are-in-preview) for details on selecting preview rules.
```

I'd also consider putting it in a note admonition.

> Simply enabling preview mode does not imply that all the new preview features/changes will be turned on. Some, like the rules that are in preview, need to be turned on explicitly (see the [rules section](#using-rules-that-are-in-preview). Others, like the formatter style or default values (for example `output-format`) will be turned on by simply enabling the preview mode.

Reviewer 2: Nit: I would avoid explicitly naming `output-format` here. We otherwise have to come up with a new preview feature everytime we remove the feature mentioned in the documentation.

Author: I hesitated when writing it, better keep it general indeed.
Removed in b0a9db688.

    preview = true
    ```

=== "CLI"

    ```console
    ruff format . --preview
    ```

## Using rules that are in preview

If a rule is marked as preview, it can only be selected if preview mode is enabled. For example, consider a
hypothetical rule, `HYP001`. If `HYP001` were in preview, it would _not_ be enabled by adding following to your

> config file or only using the `--select` flag on the CLI:

Reviewer: If you want to add more here, I'd rephrase entirely

> If `HYP001` were in preview, it would _not_ be enabled by adding it to the selected rules.

Author: Done in 9b6b7bd19.

    extend-select = ["HYP001"]
    ```

=== "CLI"

    ```console

>     ruff check . --select HYP001

Reviewer: The `.` is superfluous in all of these examples, that's the default behavior these days.

Reviewer: We're using `extend-select` in all of the other examples, why use `--select` here?

Author: > The . is superfluous in all of these examples, that's the default behavior these days.

I chose to use the `.` to be consistent with the rest of the documentation (for example, see [the tutorial page](https://docs.astral.sh/ruff/tutorial/)). If it's better to remove the `.`, then I can open another PR to remove it from the rest of the documentation, what do you think ?

> We're using extend-select in all of the other examples, why use --select here?

It was a mistake, I fixed it in 9f5d9bcdf.

Reviewer: Sure we can address that separately, thanks!

Author: Ok, I'll do that then. I've removed the `.` in this commit [356c4b0](https://github.com/astral-sh/ruff/pull/10168/commits/356c4b0580f63a918936a04b061b6503a28470d0).

    select = ["ALL"]
    ```

=== "CLI"

    ```console
    ruff check . --select ALL
    ```

> However, it would be enabled in any of the above cases if you enabled preview in your configuration file or provided the `--preview` CLI flag:

Reviewer: Similarly, I'd just generalize these e.g.

```suggestion
However, it would be enabled in any of the above cases if you enabled preview mode:
```
